<a href="https://colab.research.google.com/github/SamShinwari/Advanced-AI-Bootcamp-2026/blob/main/Simple_Neural_Networks.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Part 15: Building a Simple Neural Network

### Input

We define the input features for a single sample:
- Age = 20
- Experience = 2

This is represented as a column vector `X`:

In [1]:
import numpy as np

X = np.array([
    [20],
    [2]
])

print("Input vector X:")
print(X)

Input vector X:
[[20]
 [ 2]]


### Weights

These are the weights `W` connecting the input layer to the hidden layer (or output layer in this simple case). Each row corresponds to a neuron, and each column corresponds to an input feature.

In [2]:
W = np.array([
    [0.4, 0.8],
    [0.7, 0.2]
])

print("Weight matrix W:")
print(W)

Weight matrix W:
[[0.4 0.8]
 [0.7 0.2]]


### Bias

The bias `b` is added to the weighted sum of inputs for each neuron.

In [3]:
b = np.array([
    [1],
    [1]
])

print("Bias vector b:")
print(b)

Bias vector b:
[[1]
 [1]]


### Neural Network Computation

The neural network computes the output `Y` using the formula:
`Y = WX + b`

In [4]:
Y = W @ X + b

print("Computed output Y:")
print(Y)

Computed output Y:
[[10.6]
 [15.4]]


### Step-by-Step Calculation

Let's break down the calculation for each neuron:

**First neuron:**
- `(0.4 × 20) + (0.8 × 2) = 8 + 1.6 = 9.6`
- `9.6 + bias (1) = 10.6`

**Second neuron:**
- `(0.7 × 20) + (0.2 × 2) = 14 + 0.4 = 14.4`
- `14.4 + bias (1) = 15.4`

### Final Output

Based on the calculations, the final output `Y` is:

In [5]:
print("Final Output Y:")
print(Y)

Final Output Y:
[[10.6]
 [15.4]]


## Part 16: ReLU Activation Function

Neural networks often apply an activation function after the weighted sum. One of the most common is the Rectified Linear Unit (ReLU).

### Formula:

`ReLU(x) = max(0, x)`

This function keeps positive values unchanged and replaces negative values with zero.

### Example Input:

Consider the following input vector `Y`:

In [6]:
import numpy as np

Y_relu_input = np.array([
    [-3],
    [5],
    [-1],
    [8]
])

print("Input for ReLU function:")
print(Y_relu_input)

Input for ReLU function:
[[-3]
 [ 5]
 [-1]
 [ 8]]


### ReLU Computation:

We apply the ReLU function to the input vector. This involves taking the maximum of 0 and each element in the input.

In [7]:
relu_output = np.maximum(0, Y_relu_input)

print("Output after applying ReLU:")
print(relu_output)

Output after applying ReLU:
[[0]
 [5]
 [0]
 [8]]


### Final Output:

The final output after applying the ReLU activation function is:

In [8]:
print("Final ReLU Output:")
print(relu_output)

Final ReLU Output:
[[0]
 [5]
 [0]
 [8]]
